examples/upgradeable-contract/upgradeable-flipper/upgradeable.rs

```diff
@@ -0,0 +1,191 @@
+        Upgradeable::new(T::pull_spread(ptr))
+    }
+
+    fn push_spread(&self, ptr: &mut KeyPtr) {
+        T::push_spread(&self.inner, ptr)
+    }
+
+    fn clear_spread(&self, ptr: &mut KeyPtr) {
+        T::clear_spread(&self.inner, ptr)
+    }
+}
+
+/// It is implementation of `SpreadLayout` that initialize the inner type if it is not initialized.
+impl<T: PackedLayout + SpreadAllocate> SpreadLayout for Upgradeable<T, NotInitialized> {
+    const FOOTPRINT: u64 = <T as SpreadLayout>::FOOTPRINT;
+    const REQUIRES_DEEP_CLEAN_UP: bool = <T as SpreadLayout>::REQUIRES_DEEP_CLEAN_UP;
+
+    fn pull_spread(ptr: &mut KeyPtr) -> Self {
+        if ink_env::get_contract_storage::<T>(ptr.advance_by(0))
+            .expect("could not properly decode storage entry")
+            .is_none()
+        {
+            <Self as SpreadAllocate>::allocate_spread(ptr)
+        } else {
+            Upgradeable::new(<T as SpreadLayout>::pull_spread(ptr))
+        }
+    }
+
+    fn push_spread(&self, ptr: &mut KeyPtr) {
+        <T as SpreadLayout>::push_spread(&self.inner, ptr)
+    }
+
+    fn clear_spread(&self, ptr: &mut KeyPtr) {
+        <T as SpreadLayout>::clear_spread(&self.inner, ptr)
+    }
+}
+
+/// Below the boilerplate code to implement `PackedLayout`, `SpreadAllocate`, `PackedAllocate`.
+
+impl<T: PackedLayout> PackedLayout for Upgradeable<T, Initialized> {
+    fn pull_packed(&mut self, at: &Key) {
+        <T as PackedLayout>::pull_packed(&mut self.inner, at)
+    }
+
+    fn push_packed(&self, at: &Key) {
+        <T as PackedLayout>::push_packed(&self.inner, at)
+    }
+
+    fn clear_packed(&self, at: &Key) {
+        <T as PackedLayout>::clear_packed(&self.inner, at)
+    }
+}
+
+impl<T: PackedLayout + SpreadAllocate> PackedLayout for Upgradeable<T, NotInitialized> {
+    fn pull_packed(&mut self, at: &Key) {
+        <T as PackedLayout>::pull_packed(&mut self.inner, at)
+    }
+
+    fn push_packed(&self, at: &Key) {
+        <T as PackedLayout>::push_packed(&self.inner, at)
+    }
+
+    fn clear_packed(&self, at: &Key) {
+        <T as PackedLayout>::clear_packed(&self.inner, at)
+    }
+}
+
+impl<T: SpreadAllocate + PackedLayout> SpreadAllocate for Upgradeable<T, Initialized> {
+    fn allocate_spread(ptr: &mut KeyPtr) -> Self {
+        Upgradeable::new(<T as SpreadAllocate>::allocate_spread(ptr))
+    }
+}
+
+impl<T: SpreadAllocate + PackedLayout> SpreadAllocate for Upgradeable<T, NotInitialized> {
+    fn allocate_spread(ptr: &mut KeyPtr) -> Self {
+        Upgradeable::new(<T as SpreadAllocate>::allocate_spread(ptr))
+    }
+}
+
+impl<T: PackedAllocate> PackedAllocate for Upgradeable<T, Initialized> {
+    fn allocate_packed(&mut self, at: &Key) {
+        <T as PackedAllocate>::allocate_packed(&mut self.inner, at)
+    }
+}
+
+impl<T: PackedAllocate> PackedAllocate for Upgradeable<T, NotInitialized> {
+    fn allocate_packed(&mut self, at: &Key) {
+        <T as PackedAllocate>::allocate_packed(&mut self.inner, at)
+    }
+}
```

Comment:
Can we not make these generic over `Upgradeable<T, S>` the implementations are the same for `Initialized` and `NotInitialized`?

Reply:
No, we can't, because `SpreadAllocate` requires `SpreadAllocate`. `SpreadAllocate` is implemented only for `Upgradeable<T, NotInitialized>` and `Upgradeable<T, Initialized>` so we can't create a generic implementation.